packages/controllers/src/ranks.rs

```diff
@@ -0,0 +1,257 @@
+use std::fmt;
+use thiserror::Error;
+
+use cosmwasm_std::{
+    attr, Addr, CustomQuery, Deps, DepsMut, MessageInfo, Response, StdError, StdResult,
+};
+use cw_storage_plus::Map;
+
+/// Returned from RANKS.query_rank()
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct RankResponse {
+    pub address_rank: Option<u8>, //more than 255 ranks seems impractical
+}
+
+/// Errors returned from RANKS
+#[derive(Error, Debug, PartialEq)]
+pub enum RanksError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("Caller has no rank")]
+    NoRank {},
+
+    #[error("Caller rank too low")]
+    RankTooLow {},
+
+    #[error("Caller cannot change their own rank")]
+    CannotSetOwnRank {},
+}
+
+// state/logic
+// an IndexedMap might make sense later so that admins of a given level can be easily pulled
+pub struct Ranks<'a>(Map<'a, &'a str, u8>);
+
+// this is the core business logic we expose
+impl<'a> Ranks<'a> {
+    pub const fn new(namespace: &'a str) -> Self {
+        Ranks(Map::new(namespace))
+    }
+
+    pub fn set<Q: CustomQuery>(
+        &self,
+        deps: DepsMut<Q>,
+        address: Addr,
+        address_rank: u8,
+    ) -> StdResult<()> {
+        self.0
+            .save(deps.storage, &address.to_string(), &address_rank)
+    }
+
+    pub fn get<Q: CustomQuery>(&self, deps: Deps<Q>, address: Addr) -> StdResult<Option<u8>> {
+        match self.0.load(deps.storage, &address.to_string()) {
+            Ok(val) => Ok(Some(val)),
+            Err(StdError::NotFound { .. }) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Returns Ok(true) if this is an address of the indicated rank or higher,
+    /// Ok(false) if not, and an Error if we hit an error with Api or Storage usage
+    pub fn is_rank<Q: CustomQuery>(
+        &self,
+        deps: Deps<Q>,
+        caller: &Addr,
+        required_rank: u8,
+    ) -> StdResult<bool> {
+        match self.0.load(deps.storage, &caller.to_string()) {
+            Ok(rank) => Ok(rank >= required_rank),
+            Err(StdError::NotFound { .. }) => Ok(false),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Like is_rank but returns RanksError::RankTooLow if rank is <= required_rank.
+    /// Helper for a nice one-line auth check.
+    pub fn assert_rank<Q: CustomQuery>(
+        &self,
+        deps: Deps<Q>,
+        caller: &Addr,
+        required_rank: u8,
+    ) -> Result<(), RanksError> {
+        if !self.is_rank(deps, caller, required_rank)? {
+            Err(RanksError::RankTooLow {})
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn execute_update_rank<C, Q: CustomQuery>(
+        &self,
+        deps: DepsMut<Q>,
+        info: MessageInfo,
+        target_address: Addr,
+        new_rank: u8,
+    ) -> Result<Response<C>, RanksError>
+    where
+        C: Clone + fmt::Debug + PartialEq + JsonSchema,
+    {
+        // only addresses of >= rank than the target rank can access this
```

Comment:
are you sure addresses that = rank can call this?
can caller update rank above their rank?

alice is rank 3 
bob is at rank 99 
can alice update rank of bob to 2?